gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -131,16 +133,34 @@ OGRFlatGeobufLayer::OGRFlatGeobufLayer(
     const char *pszFilename,
     OGRSpatialReference *poSpatialRef,
     OGRwkbGeometryType eGType,
-    VSILFILE *poFpWrite,
-    const std::string& oTempFile,
-    bool bCreateSpatialIndexAtClose) :
-    m_eGType(eGType),
-    m_poFpWrite(poFpWrite),
-    m_oTempFile(oTempFile)
+    bool bCreateSpatialIndexAtClose,
+    char **papszOptions) :
+    m_eGType(eGType)
 {
     m_create = true;
     m_bCreateSpatialIndexAtClose = bCreateSpatialIndexAtClose;
 
+    int savedErrno;
+    if (bCreateSpatialIndexAtClose) {
+        CPLDebug("FlatGeobuf", "Spatial index requested will write to temp file and do second pass on close");
+        m_oTempFile = GetTempFilePath(pszFilename, papszOptions);
+        m_poFpWrite = VSIFOpenL(m_oTempFile.c_str(), "w+b");
+        savedErrno = errno;
+        // Unlink it now to avoid stale temporary file if killing the process
+        // (only works on Unix)
+        VSIUnlink(m_oTempFile.c_str());
+    } else {
+        CPLDebug("FlatGeobuf", "No spatial index will write directly to output");
+        m_poFpWrite = VSIFOpenL(pszFilename, "wb");
+        savedErrno = errno;
+    }
+    if (m_poFpWrite == nullptr) {
+        CPLError(CE_Failure, CPLE_OpenFailed,
```

Comment:
a constructor that fails is not a good thing. All that logic that can fail should likely be in a bool Initialize() method whose caller would test the return code

Reply:
even if I break out that logic, how would I handle the case differently? I cannot construct in this case.. should I avoid CPLError anyway?

Reply:
or are you saying I should revert it back to the behavior in that the poFpWrite and oTempFile should be created ahead of construction and injected?

Reply:
If you want to do minimal changes, you could have a IsValid() method that callers of the constructor should call to check if the object is valid.
Or perhaps better, make the constructor private, and have a static Create() method that returns a pointer to it or a nullptr in case of error.